File: IsraelHiking.Web/android/app/src/main/java/com/mapeak/car/CarPlugin.java
package com.mapeak.car;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "Car")
public class CarPlugin extends Plugin implements CarMessageBus.CarEventListener {

    private boolean isConnected = false;

    @Override
    public void load() {
        CarMessageBus.getInstance().registerListener(this);
    }

    @Override
    protected void handleOnDestroy() {
        CarMessageBus.getInstance().unregisterListener(this);
        super.handleOnDestroy();
    }

    @PluginMethod
    public void sendMessage(PluginCall call) {
        String type = call.getString("type");
        if (type == null || type.trim().isEmpty()) {
            call.reject("type is required");
            return;
        }

        var payload = call.getObject("payload");
        CarMessageBus.getInstance().emitEvent(new CarMessageBus.CarEvent(type, payload));
        call.resolve();
    }

    @PluginMethod
    public void getConnectionState(PluginCall call) {
        var payload = new JSObject();
        payload.put("connected", isConnected);
        call.resolve(payload);
    }

    @Override
    public void onCarEvent(CarMessageBus.CarEvent event) {
        switch (event.actionId()) {
            case CarMessageBus.EVENT_CONNECTED:
                isConnected = event.payload().getBool("connected");
                raiseEvent(event.actionId(), event.payload());
                break;
            case CarMessageBus.EVENT_LOCATION:
                raiseEvent(event.actionId(), event.payload());
                break;
        }
    }

    private void raiseEvent(String actionId, JSObject payload) {
        if (getBridge() != null) {
            getBridge().executeOnMainThread(() -> {
                notifyListeners(actionId, payload);
            });
        }
    }
}

File: IsraelHiking.Web/android/app/src/main/java/com/mapeak/car/CarPlugin.kt
package com.mapeak.car

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONArray

@CapacitorPlugin(name = "Car")
class CarPlugin : Plugin() {
    private lateinit var store: CarStore

    override fun load() {
        store = CarStore.get(context)
    }

    @Suppress("unused")
    @PluginMethod
    fun storeValue(call: PluginCall) {
        val key = call.getString("key")
        if (key.isNullOrBlank()) {
            call.reject("key is required")
            return
        }
        val value = call.getObject("value")
        if (value == null) {
            call.reject("value is required for key=$key")
            return
        }

        when (key) {
            CarStore.KEY_STYLE -> {
                addLayeringAnchorTo(value)
                store.saveStyle(value)
            }
            CarStore.KEY_ROUTE -> {
                if (value.optJSONArray("routes")?.length() == 0) {
                    store.clearRoutes()
                } else {
                    store.saveRoutes(value)
                }
            }
            CarStore.KEY_CONFIG -> store.saveConfig(value)
            else -> {
                call.reject("unknown key: $key")
                return
            }
        }
        call.resolve()
    }

    private fun addLayeringAnchorTo(style: JSObject) {
        val sources = style.optJSONObject("sources") ?: JSObject().also { style.put("sources", it) }
        val anchorData = JSObject()
            .put("type", "FeatureCollection")
            .put("features", JSONArray())
        val anchorSource = JSObject()
            .put("type", "geojson")
            .put("data", anchorData)
        sources.put(CarMapContainer.LAYERING_ANCHOR_SOURCE_ID, anchorSource)

        val layers = style.optJSONArray("layers") ?: JSONArray().also { style.put("layers", it) }
        val anchorLayer = JSObject()
            .put("id", CarMapContainer.LAYERING_ANCHOR_ID)
            .put("type", "circle")
            .put("source", CarMapContainer.LAYERING_ANCHOR_SOURCE_ID)
            .put("layout", JSObject().put("visibility", "none"))
        layers.put(anchorLayer)
    }
}
